### DISCUSSION TITLE
How to pass environment variable to an input?

### DISCUSSION THREAD

**Discussion by crazy-max** [community/community#26529] [Actions]
<p>Hi,</p>
<p>How can I pass environment variables to an input ?</p>
<pre><code>- run: echo "##[set-env name=NAME;]CrazyMax"
- uses: actions/hello-world-javascript-action@master
with:
who-to-greet: $NAME
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/11ab9f71c7c44bc337573ca20f84ef268f663579.png" data-download-href="/uploads/short-url/2wjOFnbEeE5KaMN1w165GSYryUV.png?dl=1" title="input_envvar.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/1951866/181098185-53f1130d-2d51-4be2-8dd9-debc6200e410.png" alt="input_envvar.png" data-base62-sha1="2wjOFnbEeE5KaMN1w165GSYryUV" width="690" height="422" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/11ab9f71c7c44bc337573ca20f84ef268f663579_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">input_envvar.png</span><span class="informations">778×476 14.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by chrispat** [comment, marked as answer]
<p>You currently can’t reference environment variables as inputs.  Unless you really want that particular value to be an environment variable I would suggest you set it as an output of the step and then you can reference that output in the input of subsiquent step.  </p>
<p><a href="https://help.github.com/en/articles/development-tools-for-github-actions#set-an-output-parameter-set-output" rel="nofollow noopener">https://help.github.com/en/articles/development-tools-for-github-actions#set-an-output-parameter-set-output</a></p>
<pre><code>- run: echo ::set-output name=action_fruit::strawberry
   id: fruit
- uses: actions/hello-world-javascript-action@master
   with:
     who-to-greet: ${{ steps.fruit.outputs.action_fruit }}
</code></pre>

**Reply by Pyrrha** [reply]
This is now deprecated. The correct usage, easier to read, should be:

```diff
-  - run: echo ::set-output name=action_fruit::strawberry
+  - run: echo "action_fruit=strawberry" >> "$GITHUB_OUTPUT"
     id: fruit
  - uses: actions/hello-world-javascript-action@master
     with:
       who-to-greet: ${{ steps.fruit.outputs.action_fruit }}
```

Source: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter

**Comment by crazy-max** [comment]
<p>Ok it works thanks. But small typo in your solution:</p>
<pre><code>${{ steps.fruit.output.action_fruit }}
</code></pre>
<p>should be:</p>
<pre><code>${{ steps.fruit.outputs.action_fruit }}
</code></pre>

**Comment by crazy-max** [comment]
<p>Btw is it going to be implemented with something like ${{ job.env.NAME }} maybe ?</p>

**Comment by tmhuysen** [comment]
<p>I can use global variables as input like this for example:</p>
<blockquote>
<p>env:</p>
</blockquote>
<blockquote>
<p>    FOO: “bar”</p>
</blockquote>
<blockquote>
<p>job:</p>
</blockquote>
<blockquote>
<p>    step:</p>
</blockquote>
<blockquote>
<p>     env:</p>
</blockquote>
<blockquote>
<p>       BAR: ${{ env.FOO }}</p>
</blockquote>
<p>So this is now implemented.</p>

**Comment by sandersaares** [comment]
<p>Isn’t this transform-env-to-output just an ugly workaround? I want to reference things like GITHUB_WORKSPACE in my input parameters without workarounds or hacks.</p>

**Comment by konradpabjan** [comment]
<p>You can now do something like this for passing in env variables.</p>
<p>Be aware, that env is just one of many contexts that you can use for passing in input</p>
<p><a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#contexts" rel="nofollow noopener">https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#contexts</a></p>
<p>If you want to pass in certain types of information in such as the workspace, you should use the github context</p>
<p><a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener">https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context</a></p>
<pre><code>jobs:
  example:
    runs-on: [ubuntu-latest]
    env:
      input: environment variable
    steps:
    # use env variable as input
    - uses: actions/hello-world-javascript-action@master
      with:
          who-to-greet: ${{ env.input }}

    # use input from the github context[https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context](https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context) - uses: actions/hello-world-javascript-action@master
      with:
          who-to-greet: ${{ github.repository}}

    - uses: actions/hello-world-javascript-action@master
      with:
          who-to-greet: ${{ github.ref }}
</code></pre>

**Reply by filipkrw** [reply]
Worked like a charm.

**Comment by javiercornejo** [comment]
<p>this is the best solution. Thanks <a class="mention" href="/konradpabjan">@konradpabjan</a> </p>

**Comment by maxkosty** [comment]
Might need to update this answer now that `::set-output` is being deprecated https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

**Comment by cognitiaclaeves** [comment]
I think this question can be interpreted in multiple ways. I found this question while looking for a way to reuse global environment variables, preferably in a consistent way, regardless of if it was done in actions, composite actions, workflows, or reusable workflows.

TLDR; - After defining global env vars, make them a job output through the special `$GITHUB_OUTPUT` variable via the cli. They can then be used in any workflow or action through `${{ needs.job_name.outputs.var_name }}`. Note that the cost of this consistency is that 'needs' must be used, for the job that sets the env variable.

... which is not needed when using env directly, but which can only be done for actions, and not reusable workflows.


For the sake of this explanation, here are two example code-snippets:

Here is the main workflow:

```
name: 'test-global-env-var'

on:
  push:
    branches: [ "dev-*" ]

permissions:
  actions: read
  contents: read
  security-events: write

env:
  AWS_DEFAULT_REGION: us-west-2

defaults:
  run:
    shell: bash

jobs:
  # This appears to be the only way to re-use global env vars in reusable workflows, as of 1/2023
  set-global-env-var:
    runs-on: ubuntu-latest
    outputs:
      aws_region_via_cli: ${{ steps.set-vars.outputs.aws_region }}
    steps:
      - name: set-vars-via-cli
        env:
          AWS_REGION: ${{ env.AWS_DEFAULT_REGION }}
        id: set-vars
        run: |
          echo "aws_region=${AWS_REGION}" >> $GITHUB_OUTPUT
          echo "concurrency_string=${{ format('{0}-env', github.workflow) }}" >> $GITHUB_OUTPUT

  reuse-global-var-in-actions:
    runs-on: ubuntu-latest
    needs:
      - set-global-env-var
    steps:
      - name: use-global-directly
        uses: actions/hello-world-javascript-action@master
        with:
            who-to-greet: ${{ env.AWS_DEFAULT_REGION }}

      # here env needs to be specified again for each step in each job where it is used
      # -- and then it needs to be repeated in the input for the job
      # This is too much work-around for the slight effect it gives.
      - name: use-var-by-job
        uses: actions/hello-world-javascript-action@master
        env:
          AWS_REGION: ${{ env.AWS_DEFAULT_REGION }}
        with:
            who-to-greet: ${{ env.AWS_REGION }}

      - name: reuse-global-var-consistently
        uses: actions/hello-world-javascript-action@master
        with:
            who-to-greet: ${{ needs.set-global-env-var.outputs.aws_region_via_cli }}

  reuse-global-var-in-workflow:
    needs:
      - set-global-env-var
    concurrency:
      group: ${{ needs.reuse-global-var.outputs.concurrency_string }}
      # included to demonstrate that needs works here
    uses: Company/repo/.github/workflows/test-reusable-workflow.yml@master
    with:
      env_name: ${{ env.AWS_DEFAULT_REGION }}

    # This should succeed. But it fails even before run-time like this:
    #  Invalid workflow file: .github/workflows/test-glob-env.yml#L76
    #  The workflow is not valid. .github/workflows/test-glob-env.yml (Line: 76, Col: 17): 
    #  Unrecognized named-value: 'env'. Located at position 1 within expression: env.AWS_DEFAULT_REGION
    #
    # It's silly to have to have to code additional complexity because the global env context 
    #    is not available to reusable workflows.
    # Additionally, it's misleading and confusing when it is available to actions, as above.
    # It cost me 45 minutes to hunt this down, and untold hours trying to figure out why the code
    #    didn't work before I had enough context to hunt it down.
    #    It may also be noteworthy that I originally ultimately concluded incorrectly 
    #    that the global env context was unavailable to any job inputs.
```

Here is a test reusable workflow:

```
on:
  workflow_call:
    inputs:
      env_name:
        required: true
        type: string

defaults:
  run:
    shell: bash

jobs:
  test-global-for-input:
    name: Test global for input
    runs-on: ubuntu-latest
    steps:
        - run: |
            echo "Passed input: ${{ inputs.env_name }}
```

I have three goals in taking the time to write this up:

1. The code that I wrote did not work, even though I appeared to be following what others were doing. It turned out that I did not have enough context about reusable workflows versus composite actions. Their syntax is similar enough that I was applying examples I was finding that were not work-flow based. I wish to save others this time and pain.
2. I came up with a solution to this, which I believe is the best solution to date, that I do not see here, and I haven't found while googling. So I'm providing my take, which works consistently.
3. I wish to make a strong case that it is silly to have to code additional complexity to overcome the limitation that appears to be arbitrarily enforced even before run-time.

As I was writing this, I finally figured out what @konradpabjan 's solution is demonstrating -- that you can use environment variables at the job level, rather than at every stage level. My examples do not include this, and I consider it outside of the scope of my purpose -- and since I now know that my mileage may vary significantly between reusable workflows, workflows and actions, I won't be doing the testing to determine if the job-level env variable works for reusable workflows. I would strongly suspect that it doesn't, but I leave this determination as an exercise for a consumer of this content.

**Comment by ron190** [comment]
As explained by @cognitiaclaeves, not being able to reference global `env` in `with` for workflow call then forces at best to duplicate the value, which looks like a maintenance problem since 2019:

```yml
env:
  DOCKER_RESOURCES: ./model/src/test/resources/docker

jobs:
  build-jdk:
    uses: ./.github/workflows/run-build-jdk.yml
    with:
      DOCKER_RESOURCES: ./model/src/test/resources/docker
```
At worst then you have to run a new container to define a reusable ouput:
```yml
env:
  DOCKER_RESOURCES: ./model/src/test/resources/docker

jobs:

  set-global-env-var:
    runs-on: ubuntu-latest
    steps:
      - id: set-vars
        env:
          DOCKER_RESOURCES: ${{ env.DOCKER_RESOURCES }}
        run: echo "DOCKER_RESOURCES=${DOCKER_RESOURCES}" >> $GITHUB_OUTPUT
    outputs:
      DOCKER_RESOURCES: ${{ steps.set-vars.outputs.DOCKER_RESOURCES }}

  build-jdk11:
    needs: set-global-env-var
    uses: ./.github/workflows/build-jdk11.yml
    with:
      DOCKER_RESOURCES: ${{ needs.set-global-env-var.outputs.DOCKER_RESOURCES }}
```
Hoping there's a feature in the pipe at GH to make it cleaner.
